File: .github/workflows/scripts/retrieve_spec_id.py
import yaml
import requests
import base64
import os


base_url = 'https://dash.readme.com/api/v1/'
version_url = '{}/version'.format(base_url)
api_key = 'v8f6zlqQ6wIUxDzULdsHoTZrBXTeLUcK'


def main():
    curr_version = retrieve_version_from_openapi()
    #print('Required version: {}'.format(curr_version))

    if curr_version:
        #print('Retrieving versions...')
        versions_list =  get_versions()
        #print('Done.')

        spec_id = None
        last_version = None

        for elem in versions_list.json():
            v = elem['version']
            if v == curr_version:
                spec_id = elem['_id']
            if not last_version or last_version < v:
                last_version = v
        
        if not spec_id:
            #print('Required version not found. Creating...')
            #print('Last version: {}'.format(last_version))
            res = create_version(curr_version, last_version).json()
            #print(res)
            #print('Created.')
            spec_id = res['_id']

        #print('Spec ID: {}'.format(spec_id))
        print('oas_key={}:{}'.format(api_key, spec_id))
    else:
        raise Exception('Version not found!!!')


def retrieve_version_from_openapi():
    version = None
    with open('./openapi.yaml', 'r') as stream:
        data_loaded = yaml.safe_load(stream)
        version = data_loaded['info']['version']
        version = version.replace('v', '')
    return version


def get_versions():
    api_key_b64 = base64.b64encode(api_key.encode('utf-8')).decode('utf-8')

    headers = {
        'Accept': 'application/json',
        'Authorization': 'Basic {}'.format(api_key_b64)
    }

    response = requests.request('GET', version_url, headers=headers)
    return response


def create_version(version, last_version):
    api_key_b64 = base64.b64encode(api_key.encode('utf-8')).decode('utf-8')

    headers = {
        'Accept': 'application/json',
        'Content-Type': 'application/json',
        'Authorization': 'Basic {}'.format(api_key_b64)
    }

    payload = {
        'version': version,
        'from': last_version,
        'is_stable': False,
        'is_hidden': True
    }
    #print(payload)

    response = requests.request('POST', version_url, headers=headers, json=payload)
    return response


main()

File: .github/workflows/scripts/retrieve-spec-id.js
const yaml = require('js-yaml')
const semver = require('semver')
const https = require('https')
const fs   = require('fs')

var baseUrl = 'dash.readme.com'
var path = '/api/v1/version'
var apiKey = process.argv[2]
var yamlFile = "../../../openapi.yaml"

function main() {
	var bumpVersion = retrieveVersionFromOpenApi(yamlFile)

	var result = {}

	if (bumpVersion != null) {

		var specId = null
		var major = semver.major(bumpVersion)
		var minor = semver.minor(bumpVersion)
		var apiVersion = `v${major}.${minor}`
		result['api_version'] = apiVersion

		getVersions(apiKey, function lv(versionList) {
			var versionList = JSON.parse(versionList)
			var lastVersion = null
			var compareVersion = semver.coerce(apiVersion).version

			for (var elem in versionList) {

				var elemVersion = versionList[elem]['version']
				var v = semver.coerce(elemVersion).version
				
				if (semver.eq(v, compareVersion)) {
					specId = versionList[elem]['_id']
				} 
				if (lastVersion == null || semver.lt(lastVersion, v)) {
					lastVersion = v
				}

			}

			if (specId == null) {

				/*createVersion(apiVersion, lastVersion, apiKey, function cv(response) {
					
					specId = JSON.parse(response)['_id']
					if (specId == null) {
						throw "SpecId not found!!!"
					} else {
						result['spec_id'] = specId
						printResult(result)
					}
				})*/
				getVersion(lastVersion, apiKey, function cv(response) {
					
					specId = JSON.parse(response)['_id']
					if (specId == null) {
						throw "SpecId not found!!!"
					} else {
						result['spec_id'] = specId
						printResult(result)
					}
				})
			}  else {
				result['spec_id'] = specId
				printResult(result)
			}
		})
		
	} else {
		throw "OpenAPI spec version not found!!!"
	}
	
}

main()

function printResult(result) {
	console.log(yaml.dump(result))
}

function retrieveVersionFromOpenApi(contents) {

	const openapi = yaml.load(fs.readFileSync(contents, 'utf8'))
	return openapi.info.version

}

function getVersions(apiKey, callback) {

	var apiKeyB64 = Buffer.from(apiKey).toString('base64')

	var headers = {
		'Accept': 'application/json',
		'Content-Type': 'application/json',
		'Authorization': 'Basic ' + apiKeyB64
	}

	const options = {
		hostname: baseUrl,
		path: path,
		headers: headers
	}

	https.get(options, (response) => {

		var result = ''
		response.on('data', function (chunk) {
			result += chunk
		})

		response.on('end', function () {
			callback(result)
		})

	})
}

function getVersion(version, apiKey, callback) {

	var apiKeyB64 = Buffer.from(apiKey).toString('base64')

	var headers = {
		'Accept': 'application/json',
		'Content-Type': 'application/json',
		'Authorization': 'Basic ' + apiKeyB64
	}

	const options = {
		hostname: baseUrl,
		path: path + '/' + version,
		headers: headers
	}

	https.get(options, (response) => {

		var result = ''
		response.on('data', function (chunk) {
			result += chunk
		})

		response.on('end', function () {
			callback(result)
		})

	})
}

function createVersion(version, lastVersion, apiKey, callback) {

	var apiKeyB64 = Buffer.from(apiKey).toString('base64')

	var headers = {
		'Accept': 'application/json',
		'Content-Type': 'application/json',
		'Authorization': 'Basic ' + apiKeyB64
	}

	var payload = {
		'version': version,
		'from': lastVersion,
		'is_stable': false,
		'is_hidden': true
	}

    var json = JSON.stringify(payload)
    
    const options = {
		hostname: baseUrl,
		path: path,
		method: 'POST',
		json: true,
		headers: headers
	}

    var req = https.request(options, (response) => {
  
        var result = ''
		response.on('data', function (chunk) {
			result += chunk
		})

		response.on('end', function () {
			callback(result)
		})
      })
      
      req.on('error', (e) => {
        throw e
      })
      
      req.write(json)
      req.end()

}